cpp/ql/src/Likely Bugs/OO/UnsafeUseOfThis.qhelp

```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+
+<overview>
+<p>This rule finds calls to pure virtual member functions in constructors and destructors. Such calls do not perform virtual dispatch, and can cause undefined behavior.</p>
```

Comment:
Can you sharpen this paragraph to make it more operational/practical? We can go beyond what the standard says (undefined behaviour) and discuss the likely security implications (program abort, leading to denial of service).

Also, I'm not convinced about the statement that these calls never perform virtual dispatch. How else do indirect virtual calls calls work (i.e., the ones that crash in practice without a compiler warning)?

Reply:
You're right. I took inspiration from the description in https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors, where they state that `the virtual call mechanism is disabled` (which I guess is true from a standards point of view, but doesn't quite match the implementation). I guess a more precise way to state it would be `such calls rely on the virtual table, which has not necessarily been initialized before calling the virtual function.`

Reply:
I suspect that when people say the virtual call mechanism is disabled, that refers to an optimisation that's valid to apply to calls that come directly from the constructor. Perhaps it's even a mandatory optimisation leading to a mandatory compile failure when calling a pure virtual function directly? But virtual dispatch works just fine, AFAICT, through indirect calls. The vpointer points to a sensible and consistent vtable during construction, but it'll get overwritten to point to a different vtable after construction.

Here's an example in Compiler Explorer: https://godbolt.org/z/8hnrdM. The `rbi` and `rdx` registers are `this`, and the vpointer at `[rbi]` is written to twice during the construction of a `C`.

Reply:
4a7f9100e41d58cb7f8a96784307ce40fd2a4aee takes another stab at a (hopefully) more correct explanation.